In [95]:
!pip install requests
import json
!pip install requests pandas
!pip install requests
import requests
!pip install pandas
import pandas as pd

In [30]:
import requests
from requests.adapters import HTTPAdapter, Retry

# Set up retries for the requests session
session = requests.Session()
retries = Retry(total=5, backoff_factor=0.3, status_forcelist=[500, 502, 503, 504])
session.mount("https://", HTTPAdapter(max_retries=retries))

In [43]:
import requests
from requests.exceptions import ChunkedEncodingError, RequestException
import time

In [45]:
from requests.adapters import HTTPAdapter
from requests.packages.urllib3.util.retry import Retry

# Create a session object
session = requests.Session()

# Define retry strategy
retries = Retry(total=3, backoff_factor=1, status_forcelist=[500, 502, 503, 504])
session.mount('https://', HTTPAdapter(max_retries=retries))

response = session.get(url, headers=headers)
print(response.text)


{"message":"Invalid request token."}


In [48]:
import requests
from requests.adapters import HTTPAdapter
from requests.packages.urllib3.util.retry import Retry

# URL and headers
url = "https://api.foursquare.com/v3/places/search?ll=40.748817,-73.985428"  # Example coordinates
headers = {
    "Accept": "application/json",
    "Authorization": "Bearer fsq3TnrRSb9XQ006wG7hOHYgmzc2U4srM6sf12E7ZZ2j8RI="  # Your token
}

# Create a session object
session = requests.Session()

# Define retry strategy
retries = Retry(total=3, backoff_factor=1, status_forcelist=[500, 502, 503, 504])
session.mount('https://', HTTPAdapter(max_retries=retries))

# Make the request
response = session.get(url, headers=headers)

# Check the response
if response.status_code == 200:
    print(response.json())
else:
    print(f"Request failed with status code {response.status_code}: {response.text}")

Request failed with status code 401: {"message":"Invalid request token."}


In [46]:
import requests

url = "https://api.foursquare.com/v3/places/search"

headers = {
    "Accept": "application/json",
    "Authorization": "Bearer fsq3TnrRSb9XQ006wG7hOHYgmzc2U4srM6sf12E7ZZ2j8RI="
}

response = requests.request("GET", url, headers=headers)

print(response.text)

ChunkedEncodingError: ('Connection broken: IncompleteRead(0 bytes read, 36 more expected)', IncompleteRead(0 bytes read, 36 more expected))

In [65]:
import requests

api_key = "fsq36+J+v0B1F73kFx78DEgqNPVIa0JHVSix9VIsbOQ0OdU="

url = "https://api.foursquare.com/v3/places/search"

headers = {
    "accept": "application/json",
    "Authorization": "fsq36+J+v0B1F73kFx78DEgqNPVIa0JHVSix9VIsbOQ0OdU="
}

response = requests.get(url, headers=headers, stream=True)

try:
    # Check if the response status is OK
    response.raise_for_status()
    
    # Read the response in chunks
    for chunk in response.iter_content(chunk_size=1024):
        print(chunk.decode('utf-8'))  # Decode each chunk and print

except requests.exceptions.RequestException as e:
    print(f"Request failed: {e}")


{"results":[{"fsq_id":"4b0586f6f964a520297822e3","categories":[{"id":13034,"name":"Café","short_name":"Café","plural_name":"Cafés","icon":{"prefix":"https://ss3.4sqi.net/img/categories_v2/food/cafe_","suffix":".png"}},{"id":13065,"name":"Restaurant","short_name":"Restaurant","plural_name":"Restaurants","icon":{"prefix":"https://ss3.4sqi.net/img/categories_v2/food/default_","suffix":".png"}}],"chains":[],"closed_bucket":"LikelyOpen","distance":1315,"geocodes":{"main":{"latitude":53.524012,"longitude":-113.513161},"roof":{"latitude":53.524012,"longitude":-113.513161}},"link":"/v3/places/4b0586f6f964a520297822e3","location":{"address":"10922 88 Ave NW","country":"CA","formatted_address":"10922 88 Ave NW, Edmonton AB T6G 0Z1","locality":"Edmonton","postcode":"T6G 0Z1","region":"AB"},"name":"Sugarbowl","related_places":{},"timezone":"America/Edmonton"},{"fsq_id":"53862821498e6c94a4b173b3","categories":[{"id":13000,"name":"Dining and Drinking","short_name":"Dining and Drinking","plural_name"

# Foursquare

Send a request to Foursquare with a small radius (1000m) for all the bike stations in your city of choice. 

In [72]:
# Foursquare API credentials
FOURSQUARE_API_KEY = "fsq36+J+v0B1F73kFx78DEgqNPVIa0JHVSix9VIsbOQ0OdU="
FOURSQUARE_URL = "https://api.foursquare.com/v3/places/search"
FOURSQUARE_HEADERS = {
    "Authorization": f"{FOURSQUARE_API_KEY}", 
    "Content-Type": "application/json"
}

In [25]:

# Set search radius and query (e.g., restaurants, bars)
radius = 1000 
categories = "13065,13003"  # Foursquare category IDs: restaurants, bars

In [3]:
# Monaco bike stations data
monaco_bike_stations = [
    {"name": "Grimaldi Forum", "latitude": 43.743545, "longitude": 7.4310860000000005},
    {"name": "ÉGLISE", "latitude": 43.74495342677297, "longitude": 7.427207000000001}
]

In [4]:
# Initialize results list
foursquare_results = []

In [81]:
# Query Foursquare API
def query_foursquare(lat, lon, radius=1000, categories="13065,13003", retries=5, backoff_factor=0.5):
    params = {
        "ll": f"{lat},{lon}",
        "radius": radius,
        "categories": categories,
        "limit": 50
    }
    retries = int(retries)
    for attempt in range(retries):
        try:
            response = requests.get(FOURSQUARE_URL, headers=headers, params=params, timeout=10)
            response.raise_for_status()  # Raise an exception for HTTP errors
            return response.json()  # Return JSON response on success
        except ChunkedEncodingError as e:
            print(f"ChunkedEncodingError encountered. Retrying... ({attempt + 1}/{retries})")
            time.sleep(backoff_factor * (2 ** attempt))  # Exponential backoff
        except RequestException as e:
            print(f"Request failed for {lat}, {lon}: {e}")
            break  # Exit on non-recoverable errors

    return None  # Return None if all retries fail

In [75]:
# Make API request
params = {
        "ll": f"{43.732431},{7.419778000000001}",
        "radius": radius,
        "categories": categories,
        "limit": 50
    }
response = requests.get(FOURSQUARE_URL, params=params, headers=headers, timeout=20)
    
if response.status_code == 200:
        data = response.json()
        print(response.text)

{"results":[{"fsq_id":"4c0bf1016071a593a502e232","categories":[{"id":13148,"name":"French Restaurant","short_name":"French","plural_name":"French Restaurants","icon":{"prefix":"https://ss3.4sqi.net/img/categories_v2/food/french_","suffix":".png"}}],"chains":[],"closed_bucket":"VeryLikelyOpen","distance":300,"geocodes":{"main":{"latitude":43.732722,"longitude":7.423496}},"link":"/v3/places/4c0bf1016071a593a502e232","location":{"address":"4, quai Antoine 1er","country":"MC","formatted_address":"4, quai Antoine 1er, 98000 Monaco Monte-Carlo","locality":"Monte-Carlo","postcode":"98000 Monaco"},"name":"Quai des Artistes","related_places":{},"timezone":"Europe/Paris"},{"fsq_id":"4bf81303abdaef3b509da184","categories":[{"id":13383,"name":"Steakhouse","short_name":"Steakhouse","plural_name":"Steakhouses","icon":{"prefix":"https://ss3.4sqi.net/img/categories_v2/food/steakhouse_","suffix":".png"}}],"chains":[],"closed_bucket":"VeryLikelyOpen","distance":517,"geocodes":{"main":{"latitude":43.7279

In [83]:
test_json= query_foursquare(lat="43.732431", lon="7.419778000000001")
test_json

{'results': [{'fsq_id': '4c0bf1016071a593a502e232',
   'categories': [{'id': 13148,
     'name': 'French Restaurant',
     'short_name': 'French',
     'plural_name': 'French Restaurants',
     'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/french_',
      'suffix': '.png'}}],
   'chains': [],
   'closed_bucket': 'VeryLikelyOpen',
   'distance': 300,
   'geocodes': {'main': {'latitude': 43.732722, 'longitude': 7.423496}},
   'link': '/v3/places/4c0bf1016071a593a502e232',
   'location': {'address': '4, quai Antoine 1er',
    'country': 'MC',
    'formatted_address': '4, quai Antoine 1er, 98000 Monaco Monte-Carlo',
    'locality': 'Monte-Carlo',
    'postcode': '98000 Monaco'},
   'name': 'Quai des Artistes',
   'related_places': {},
   'timezone': 'Europe/Paris'},
  {'fsq_id': '4bf81303abdaef3b509da184',
   'categories': [{'id': 13383,
     'name': 'Steakhouse',
     'short_name': 'Steakhouse',
     'plural_name': 'Steakhouses',
     'icon': {'prefix': 'https://ss3.4sqi

In [82]:
categories

'13065,13003'

Parse through the response to get the POI (such as restaurants, bars, etc) details you want (ratings, name, location, etc)

In [17]:
def parse_foursquare(data):
    results = []
    if data and "results" in data:
        for poi in data["results"]:
            results.append({
                "name": poi.get("name"),
                "category": poi["categories"][0]["name"] if poi.get("categories") else None,
                "address": poi["location"]["formatted_address"] if "location" in poi else None,
                "lat": poi["geocodes"]["main"]["latitude"] if "geocodes" in poi else None,
                "lon": poi["geocodes"]["main"]["longitude"] if "geocodes" in poi else None
            })
    return results

In [20]:
bike_stations = [
    {"lat": 43.732431, "lon": 7.419778000000001},  # Place d'Armes
    {"lat": 43.725515, "lon": 7.41877},  # Heliport
    {"lat": 43.727572, "lon": 7.418872000000001},  # Roseraie
    {"lat": 43.72763, "lon": 7.414385},  # Stade Louis II (P)
    {"lat": 43.729167, "lon": 7.4145},  # Gabian
    {"lat": 43.730026, "lon": 7.416254},  # Fontvieille
    {"lat": 43.72854, "lon": 7.41926},  # Port de Fontvieille
    {"lat": 43.730999694368094, "lon": 7.413016124237069},  # Parking des Salines bas
    {"lat": 43.73155012533212, "lon": 7.413498921859748},  # Parking des Salines haut
    {"lat": 43.73147260015036, "lon": 7.41695360707093}  # Collège Charles III
]

In [84]:
foursquare_results = []

for station in bike_stations:
    lat, lon = station["lat"], station["lon"]
    print(f"Querying Foursquare for station at {lat}, {lon}...")
    
    # Query Foursquare
    fs_data = query_foursquare(lat, lon)
    if fs_data:
        foursquare_results.extend(parse_foursquare(fs_data))

print(foursquare_results)

Querying Foursquare for station at 43.732431, 7.419778000000001...
Querying Foursquare for station at 43.725515, 7.41877...
Querying Foursquare for station at 43.727572, 7.418872000000001...
Querying Foursquare for station at 43.72763, 7.414385...
Querying Foursquare for station at 43.729167, 7.4145...
Querying Foursquare for station at 43.730026, 7.416254...
Querying Foursquare for station at 43.72854, 7.41926...
Querying Foursquare for station at 43.730999694368094, 7.413016124237069...
Querying Foursquare for station at 43.73155012533212, 7.413498921859748...
Querying Foursquare for station at 43.73147260015036, 7.41695360707093...
[{'name': 'Quai des Artistes', 'category': 'French Restaurant', 'address': '4, quai Antoine 1er, 98000 Monaco Monte-Carlo', 'lat': 43.732722, 'lon': 7.423496}, {'name': 'BeefBar', 'category': 'Steakhouse', 'address': '42 quai Jean-Charles Rey, Fontvieille', 'lat': 43.727966, 'lon': 7.421609}, {'name': 'La Rascasse', 'category': 'Lounge', 'address': '1 qua

Put your parsed results into a DataFrame

In [85]:
# Create Foursquare DataFrame
foursquare_df = pd.DataFrame(foursquare_results)
print("Foursquare DataFrame:")
print(foursquare_df.head())

Foursquare DataFrame:
                name             category  \
0  Quai des Artistes    French Restaurant   
1            BeefBar           Steakhouse   
2        La Rascasse               Lounge   
3         La Saliere   Italian Restaurant   
4        Moshi Moshi  Japanese Restaurant   

                                         address        lat       lon  
0  4, quai Antoine 1er, 98000 Monaco Monte-Carlo  43.732722  7.423496  
1          42 quai Jean-Charles Rey, Fontvieille  43.727966  7.421609  
2         1 quai Antoine 1er, 98000 Monaco-Ville  43.732656  7.423194  
3   28 Quai Jean-Charles Rey, 98000 Monaco-Ville  43.728842  7.420552  
4                                                 43.728590  7.420829  


# Yelp

Send a request to Yelp with a small radius (1000m) for all the bike stations in your city of choice. 

In [86]:
YELP_URL = "https://api.yelp.com/v3/businesses/search"
YELP_HEADERS = {
    "Authorization": f"Bearer 7pRX9ejswtlgdQdtWHsfqOvENtUtrNi-tEwvwCWDJbTy9KzPqOTcbbinGZPd3MONVbCjp0VtxLGJVRYpv_ME5biD9a3X4-wOsoPUZMNR2a-tk2xkaSOG4-G8wPhgZ3Yx"
}

In [87]:
def query_yelp(lat, lon, radius=1000, categories="restaurants,bars"):
    params = {
        "latitude": lat,
        "longitude": lon,
        "radius": radius,
        "categories": categories,
        "limit": 50
    }
    response = requests.get(YELP_URL, headers=YELP_HEADERS, params=params)
    if response.status_code == 200:
        return response.json()
    else:
        print(f"Yelp API Error: {response.status_code}")
        return None

Parse through the response to get the POI (such as restaurants, bars, etc) details you want (ratings, name, location, etc)

In [88]:
def parse_yelp(data):
    results = []
    if data and "businesses" in data:
        for business in data["businesses"]:
            results.append({
                "name": business.get("name"),
                "category": business["categories"][0]["title"] if business.get("categories") else None,
                "address": ", ".join(business["location"]["display_address"]),
                "rating": business.get("rating"),
                "lat": business["coordinates"]["latitude"] if "coordinates" in business else None,
                "lon": business["coordinates"]["longitude"] if "coordinates" in business else None
            })
    return results

In [90]:
yelp_results = []

for station in bike_stations:
    lat, lon = station["lat"], station["lon"]
    
    # Query Yelp
    yelp_data = query_yelp(lat, lon)
    yelp_results.extend(parse_yelp(yelp_data))
print(yelp_results)

[{'name': 'Planet Sushi', 'category': 'Sushi Bars', 'address': '42 bd Jean Jaurès, 06300 Nice, France', 'rating': 2.8, 'lat': 43.7346103246793, 'lon': 7.42137582104865}, {'name': 'New Indian', 'category': 'Indian', 'address': '9 bd République, 06240 Beausoleil, France', 'rating': 4.0, 'lat': 43.74253, 'lon': 7.42527}, {'name': 'La Terrasse', 'category': 'Italian', 'address': 'Avenue St Martin, 06320 Monaco, France', 'rating': 4.0, 'lat': 43.7305826816304, 'lon': 7.42547103409965}, {'name': "A'Trego", 'category': 'Mediterranean', 'address': "Port de Cap d'Ail, 06320 Cap d'Ail, France", 'rating': 3.5, 'lat': 43.72505, 'lon': 7.41346}, {'name': 'Babek Kebab', 'category': 'Kebab', 'address': 'Laredo, TX 78045', 'rating': 0.0, 'lat': 43.73384990926642, 'lon': 7.421129595751689}, {'name': 'Babek Kebab', 'category': 'Fast Food', 'address': 'Laredo, TX 78045', 'rating': 0.0, 'lat': 43.73384990926642, 'lon': 7.421129595751689}, {'name': 'Babek Kebab', 'category': 'Kebab', 'address': 'Laredo, TX

Put your parsed results into a DataFrame

In [91]:
yelp_df = pd.DataFrame(yelp_results)
print("\nYelp DataFrame:")
print(yelp_df.head())


Yelp DataFrame:
           name       category                                     address  \
0  Planet Sushi     Sushi Bars       42 bd Jean Jaurès, 06300 Nice, France   
1    New Indian         Indian   9 bd République, 06240 Beausoleil, France   
2   La Terrasse        Italian      Avenue St Martin, 06320 Monaco, France   
3       A'Trego  Mediterranean  Port de Cap d'Ail, 06320 Cap d'Ail, France   
4   Babek Kebab          Kebab                            Laredo, TX 78045   

   rating        lat       lon  
0     2.8  43.734610  7.421376  
1     4.0  43.742530  7.425270  
2     4.0  43.730583  7.425471  
3     3.5  43.725050  7.413460  
4     0.0  43.733850  7.421130  


In [ ]:
# save df as CSV file
yelp_df.to_csv("yelp_df.csv", index=False)

# Comparing Results

Which API provided you with more complete data? Provide an explanation. 

In [93]:
print("Foursquare POI Count:", len(foursquare_df))
print("Yelp POI Count:", len(yelp_df))


Foursquare POI Count: 500
Yelp POI Count: 418


The Foursquare API provided more complete data than the Yelp API, as we can see by the higher Point of Interest (POI) count (500 compared to 418). This indicates that Foursquare captured a broader range of locations within the queried area. Along with this, Foursquare's results appear more consistent, with less repetitive or redundant entries compared to Yelp's results, which include multiple duplicates (ex., "Babek Kebab"). This suggests that Foursquare not only provided a more extensive dataset but also provided better data quality overall.

Get the top 10 restaurants according to their rating

In [92]:
top_10_restaurants = yelp_df.sort_values(by="rating", ascending=False).head(10)
print("\nTop 10 Restaurants by Rating (Yelp):")
print(top_10_restaurants)


Top 10 Restaurants by Rating (Yelp):
             name      category  \
374         Le 16   Restaurants   
141  Les Oliviers  Coffee & Tea   
243    New Saigon    Vietnamese   
390    Docks Café     Fast Food   
238  Les Oliviers  Coffee & Tea   
148    New Saigon    Vietnamese   
384   Indian Star        Indian   
150    Docks Café     Fast Food   
152   Indian Star        Indian   
203    Docks Café     Fast Food   

                                               address  rating        lat  \
374     16 ave du 3 Septembre, 06320 Cap d'Ail, France     5.0  43.721600   
141         Port de Cap d'ail, 06320 Cap-d'Ail, France     5.0  43.724854   
243  4 boulevard des Moneghetti, 06240 Beausoleil, ...     5.0  43.738730   
390            Port Cap D'ail, 06320 Cap D'Ail, France     5.0  43.725050   
238         Port de Cap d'ail, 06320 Cap-d'Ail, France     5.0  43.724854   
148  4 boulevard des Moneghetti, 06240 Beausoleil, ...     5.0  43.738730   
384  23 Boulevard Republique Rd53, 06